consensus/src/common.rs

```diff
@@ -75,7 +96,271 @@ impl Chain<ProtocolState> for ProtocolStateChain {
         })
     }
 
+    // FIXME: this currently returns a blake2b hash. It should
+    // return a poseidon hash according to: https://github.com/MinaProtocol/mina/blob/32f529d8d8d712a44ee75be66061ce08cbdc8924/docs/specs/consensus/README.md#517-hashstate
     fn state_hash(&self) -> Option<StateHash> {
         self.top().map(|s| s.hash())
     }
 }
+
+trait Consensus {
+    type Chain;
+    fn select_secure_chain<'a>(
+        &'a self,
+        candidates: &'a [Self::Chain],
+        constants: &ConsensusConstants,
+    ) -> Result<&'a ProtocolStateChain, ConsensusError>;
+
+    fn select_longer_chain<'a>(
+        &'a self,
+        candidate: &'a ProtocolStateChain,
+    ) -> Result<&'a ProtocolStateChain, ConsensusError>;
+}
+
```

Comment:
Excellent work on design trait Consensus  and impl Consensus for ProtocolStateChain 👍